### PR TITLE
FEATURE: Allow more than one authentication provider

### PR DESCRIPTION
**Pull request by theilm**
**What I did**

Currently, only one authentication provider can be used with Neos as described in #940. I implemented the changes proposed by the issue creator to allow more than one authentication provider.

Although I tested it successfully, I might not be aware of all possible implications; it would be great if someone with a deeper insight into the security layer / user handling of Neos could have a look.

Please note: This PR does not cover the issues that arise when more than one account is associated with a single Neos user. It only allows one account per user (a limitation that already exists right now, though).

**How I did it**

Instead of always using the default authentication provider (Neos.Neos:Backend) to retrieve a user or username, the authentication provider associated with the respective account is used (also see #940).

**Questions**

I am not sure if it should be considered a bugfix. Please let me know what you think, I could change the target branch accordingly in a new PR.

### CODE REVIEW

File: Neos.Neos/Classes/Service/UserService.php [hunk 1]
```diff
@@ -89,7 +95,10 @@ public function getPersonalWorkspaceName()
             return null;
         }
 
-        $username = $this->userDomainService->getUsername($currentUser);
+        $username = $this->userDomainService->getUsername(
+            $currentUser,
+            $this->securityContext->getAccount()->getAuthenticationProviderName()
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Service/UserService.php]: `Security\Context::getAccount()` just returns the first account that is authenticated. That is not super "deterministic". is it really what we need here?

Comment (theilm) [reply Neos.Neos/Classes/Service/UserService.php]: Do you mean the case that a single Neos user has more that one account associated? Handling this is not covered by the PR. Or are there other scenarios with more than one authenticated account? Thanks for your feedback!

Comment (bwaidelich) [reply Neos.Neos/Classes/Service/UserService.php]: See [doc comment ](https://github.com/neos/flow-development-collection/blob/master/Neos.Flow/Classes/Security/Context.php#L458-L460)on the function:
> There might be a more currently authenticated account in the remaining tokens. If you need them you'll have to fetch them directly from the tokens.

Many accounts can be authenticated with one request

Comment (theilm) [reply Neos.Neos/Classes/Service/UserService.php]: I am not sure if I understand all implications correctly. But on the other hand, the method that retrieves the $currentUser uses the same functionality:
`Domain/Service/UserService::getCurrentUser` calls `$this->securityContext->getAccount()` - so the current user should be related to the same account as in the code fragment in question.
This PR does not cover the problem, that one user could have multiple accounts / different authentication providers. Should be possible, too, but a lot more complex I guess.
What do you think? Thanks for your feedback!

Comment (bwaidelich) [reply Neos.Neos/Classes/Service/UserService.php]: > But on the other hand, the method that retrieves the $currentUser uses the same functionality

Good call! I think that, too, should use `$this->securityContext->getAccountByAuthenticationProviderName($this->defaultAuthenticationProviderName)` (or rather allow the $authProviderName to be specified).
I don't understand all the implications of this and maybe it's really a separate concern to fix. I just stumbled upon this line in your PR

**Comment by gerdemann** [comment]
I don't think this will work if the same AccountIdentifier exists in both providers. The workspace name is formed only by the AccountIdentifier.
@theilm Have you tested this case?

**Comment by theilm** [comment]
@gerdemann Thanks for pointing that out. I tested it and had a look into the code, it's true that it does not work when the same AccountIdentifier is used twice.
I am not sure it that is a blocker for the PR - at the moment it does not work at all. But it could be confusing.
By the way: I noticed that already now we run into exceptions when similar AccountIdentifiers (e.g. "test-user" and "testuser" are used, as the slugifyUsername that creates the personal workspace name creates the same workspace name for both AccountIdentifier variants (event when using only one authentication provider). I will create an issue for that (if it does not exist already).

**Comment by gjwnc** [comment]
@theilm @kitsunet Would it be possible to get this commit in the 4.3 branch too?

I try to get SAML authentication working and this commit causes the login to work.

**Comment by kitsunet** [comment]
@gjwnc this is a feature request which can by definition only go into master / the next release. Given the nature of the change I would rather not want to introduce that in a bugfix version.

**Comment by albe** [comment]
> I am not sure it that is a blocker for the PR - at the moment it does not work at all.

IMO not a blocker, since this still improves on current behaviour. Needs to be tackled together with #2850 then
So other than the test failures, this looks good - can someone take care of those? Looks like a missing mock injection

**Comment by albe** [comment]
Last call - if someone fixes the tests until tomorrow I'll merge this for the upcoming release :)

**Comment by albe** [comment]
Thanks a lot! ❤️

**Pull request merged**